File: DevExpress-Demo/DevExpress.SalesDemo.Wpf/CS/DevExpress.SalesDemo.Wpf/ViewModel/PieBarRangeViewModel/RangeViewModel.cs
using DevExpress.Mvvm.POCO;
using DevExpress.SalesDemo.Model;
using System;

namespace DevExpress.SalesDemo.Wpf.ViewModel {
    public class RangeViewModel : DataViewModel {
        public static RangeViewModel Create() {
            return ViewModelSource.Create(() => new RangeViewModel());
        }
        protected RangeViewModel() {
            AreaSeriesArgumentDataMember = "StartOfPeriod";
            AreaSeriesValueDataMember = "TotalCost";

            DateTime today = DateTime.Today;
            RangeEnd = today;
            RangeStart = today.AddDays(-DaysInMonth);
            SelectedRangeStart = RangeStart.Value.AddDays(OneThirdOfMonth);
            SelectedRangeEnd = RangeEnd.Value.AddDays(-OneThirdOfMonth);

            if(this.IsInDesignMode())
                OnInitializeInDesignMode();
        }

        const int OneThirdOfMonth = 10;
        const int DaysInMonth = 30;
        public virtual object AreaSeriesDataSource { get; protected set; }
        public virtual DateTime? RangeStart { get; set; }
        public virtual DateTime? RangeEnd { get; set; }
        public virtual DateTime? VisibleRangeStart { get; set; }
        public virtual DateTime? VisibleRangeEnd { get; set; }
        public virtual DateTime? SelectedRangeStart { get; set; }
        public virtual DateTime? SelectedRangeEnd { get; set; }
        public virtual string AreaSeriesArgumentDataMember { get; protected set; }
        public virtual string AreaSeriesValueDataMember { get; protected set; }
        public event EventHandler RangeChanged;
        public event EventHandler SelectedRangeChanged;


        public DateTimeRange? GetSelectedRange() {
            if(SelectedRangeStart == null || SelectedRangeEnd == null || SelectedRangeStart >= SelectedRangeEnd)
                return null;
            return new DateTimeRange(SelectedRangeStart.Value, SelectedRangeEnd.Value);
        }
        public void LoadData(object data) {
            AreaSeriesDataSource = data;
        }
        public void SetPrePeriod() {
            if(RangeStart == null || RangeEnd == null)
                return;
            DateTime oldRangeEnd = RangeEnd.Value;
            DateTime dateInPreMotnth = oldRangeEnd.AddMonths(-1);
            DateTime start = new DateTime(dateInPreMotnth.Year, dateInPreMotnth.Month, 1);
            int daysInPreMonts = DateTime.DaysInMonth(dateInPreMotnth.Year, dateInPreMotnth.Month);
            DateTime end = new DateTime(dateInPreMotnth.Year, dateInPreMotnth.Month, daysInPreMonts);
            RangeStart = start;
            RangeEnd = end;
        }
        public void SetNextPeriod() {
            if(RangeStart == null || RangeEnd == null)
                return;
            RangeEnd = RangeEnd.Value.AddMonths(1);
            RangeStart = RangeStart.Value.AddMonths(1);
            SetSelectedRangeInMiddle();
        }
        public bool CanSetNextPeriod() {
            return DateTime.Today > RangeEnd;
        }
        protected void OnRangeStartChanged() {
            SetSelectedRangeInMiddle();
            VisibleRangeStart = RangeStart;
            RaiseRangeChanged();
        }
        protected void OnRangeEndChanged() {
            SetSelectedRangeInMiddle();
            VisibleRangeEnd = RangeEnd;
            RaiseRangeChanged();
        }
        protected void OnSelectedRangeStartChanged() {
            RaiseSelectedRangeChanged();
        }
        protected void OnSelectedRangeEndChanged() {
            RaiseSelectedRangeChanged();
        }
        void OnInitializeInDesignMode() {
            RequestData("Data", x => x.GetSales(RangeStart.Value, RangeEnd.Value.AddDays(1), GroupingPeriod.Day), x => LoadData(x));
        }
        void SetSelectedRangeInMiddle() {
            if(RangeStart == null || RangeEnd == null || RangeStart > RangeEnd)
                return;
            SelectedRangeEnd = RangeEnd.Value.AddDays(-OneThirdOfMonth);
            SelectedRangeStart = RangeStart.Value.AddDays(OneThirdOfMonth);
        }
        void RaiseRangeChanged() {
            if(RangeChanged != null)
                RangeChanged(this, EventArgs.Empty);
        }
        void RaiseSelectedRangeChanged() {
            if(SelectedRangeChanged != null)
                SelectedRangeChanged(this, EventArgs.Empty);
        }
    }
}

File: DevExpress-Demo/DevExpress.SalesDemo.Wpf/VB/DevExpress.SalesDemo.Wpf/ViewModel/PieBarRangeViewModel/RangeViewModel.vb
Imports DevExpress.Mvvm.POCO
Imports DevExpress.SalesDemo.Model
Imports System

Namespace DevExpress.SalesDemo.Wpf.ViewModel

    Public Class RangeViewModel
        Inherits DataViewModel

        Public Shared Function Create() As RangeViewModel
            Return ViewModelSource.Create(Function() New RangeViewModel())
        End Function

        Protected Sub New()
            AreaSeriesArgumentDataMember = "StartOfPeriod"
            AreaSeriesValueDataMember = "TotalCost"
            Dim today As Date = Date.Today
            RangeEnd = today
            RangeStart = today.AddDays(-DaysInMonth)
            SelectedRangeStart = RangeStart.Value.AddDays(OneThirdOfMonth)
            SelectedRangeEnd = RangeEnd.Value.AddDays(-OneThirdOfMonth)
            If IsInDesignMode() Then OnInitializeInDesignMode()
        End Sub

        Const OneThirdOfMonth As Integer = 10

        Const DaysInMonth As Integer = 30

        Public Overridable Property AreaSeriesDataSource As Object

        Public Overridable Property RangeStart As Date?

        Public Overridable Property RangeEnd As Date?

        Public Overridable Property VisibleRangeStart As Date?

        Public Overridable Property VisibleRangeEnd As Date?

        Public Overridable Property SelectedRangeStart As Date?

        Public Overridable Property SelectedRangeEnd As Date?

        Public Overridable Property AreaSeriesArgumentDataMember As String

        Public Overridable Property AreaSeriesValueDataMember As String

        Public Event RangeChanged As EventHandler

        Public Event SelectedRangeChanged As EventHandler

        Public Function GetSelectedRange() As DateTimeRange?
            If SelectedRangeStart Is Nothing OrElse SelectedRangeEnd Is Nothing OrElse SelectedRangeStart >= SelectedRangeEnd Then Return Nothing
            Return New DateTimeRange(SelectedRangeStart.Value, SelectedRangeEnd.Value)
        End Function

        Public Sub LoadData(ByVal data As Object)
            AreaSeriesDataSource = data
        End Sub

        Public Sub SetPrePeriod()
            If RangeStart Is Nothing OrElse RangeEnd Is Nothing Then Return
            Dim oldRangeEnd As Date = RangeEnd.Value
            Dim dateInPreMotnth As Date = oldRangeEnd.AddMonths(-1)
            Dim start As Date = New DateTime(dateInPreMotnth.Year, dateInPreMotnth.Month, 1)
            Dim daysInPreMonts As Integer = Date.DaysInMonth(dateInPreMotnth.Year, dateInPreMotnth.Month)
            Dim [end] As Date = New DateTime(dateInPreMotnth.Year, dateInPreMotnth.Month, daysInPreMonts)
            RangeStart = start
            RangeEnd = [end]
        End Sub

        Public Sub SetNextPeriod()
            If RangeStart Is Nothing OrElse RangeEnd Is Nothing Then Return
            RangeEnd = RangeEnd.Value.AddMonths(1)
            RangeStart = RangeStart.Value.AddMonths(1)
            SetSelectedRangeInMiddle()
        End Sub

        Public Function CanSetNextPeriod() As Boolean
            Return Date.Today > RangeEnd
        End Function

        Protected Sub OnRangeStartChanged()
            SetSelectedRangeInMiddle()
            VisibleRangeStart = RangeStart
            RaiseRangeChanged()
        End Sub

        Protected Sub OnRangeEndChanged()
            SetSelectedRangeInMiddle()
            VisibleRangeEnd = RangeEnd
            RaiseRangeChanged()
        End Sub

        Protected Sub OnSelectedRangeStartChanged()
            RaiseSelectedRangeChanged()
        End Sub

        Protected Sub OnSelectedRangeEndChanged()
            RaiseSelectedRangeChanged()
        End Sub

        Private Sub OnInitializeInDesignMode()
            RequestData("Data", Function(x) x.GetSales(RangeStart.Value, RangeEnd.Value.AddDays(1), GroupingPeriod.Day), Sub(x) LoadData(x))
        End Sub

        Private Sub SetSelectedRangeInMiddle()
            If RangeStart Is Nothing OrElse RangeEnd Is Nothing OrElse RangeStart > RangeEnd Then Return
            SelectedRangeEnd = RangeEnd.Value.AddDays(-OneThirdOfMonth)
            SelectedRangeStart = RangeStart.Value.AddDays(OneThirdOfMonth)
        End Sub

        Private Sub RaiseRangeChanged()
            RaiseEvent RangeChanged(Me, EventArgs.Empty)
        End Sub

        Private Sub RaiseSelectedRangeChanged()
            RaiseEvent SelectedRangeChanged(Me, EventArgs.Empty)
        End Sub
    End Class
End Namespace
